File: ledgerlm/ledgerlm/server/utils/entraToken.ts
/**
 * Azure Entra ID token fetcher for PostgreSQL Managed Identity auth.
 *
 * Azure exposes two different MSI endpoints depending on the host:
 *
 *   App Service / Container Apps:
 *     URL    → $IDENTITY_ENDPOINT?resource=...&api-version=2019-08-01
 *     Header → X-IDENTITY-HEADER: $IDENTITY_HEADER
 *
 *   Bare VM / VMSS:
 *     URL    → http://169.254.169.254/metadata/identity/oauth2/token?...
 *     Header → Metadata: true
 *
 * We try the App Service path first and fall back to VM IMDS.
 * Token is cached in-process and refreshed 5 minutes before expiry.
 */

interface TokenCache {
  token:     string;
  expiresAt: number; // unix milliseconds
}

const REFRESH_BUFFER_MS = 5 * 60 * 1000;
const PG_RESOURCE  = "https://ossrdbms-aad.database.windows.net";
const COG_RESOURCE = "https://cognitiveservices.azure.com/";

// VM IMDS endpoint (fallback — not available inside App Service containers)
const VM_IMDS_URL =
  "http://169.254.169.254/metadata/identity/oauth2/token" +
  "?api-version=2019-08-01" +
  "&resource=https%3A%2F%2Fossrdbms-aad.database.windows.net";

let tokenCache: TokenCache | null = null;
let cogTokenCache: TokenCache | null = null;

export async function getEntraToken(): Promise<string> {
  const now = Date.now();

  // Return cached token if still fresh
  if (tokenCache && tokenCache.expiresAt - now > REFRESH_BUFFER_MS) {
    return tokenCache.token;
  }

  const identityEndpoint = process.env.IDENTITY_ENDPOINT;
  const identityHeader   = process.env.IDENTITY_HEADER;

  let res: Response;

  if (identityEndpoint && identityHeader) {
    // ── Azure App Service / Container Apps MSI ────────────────────────────────
    const url = `${identityEndpoint}?resource=${encodeURIComponent(PG_RESOURCE)}&api-version=2019-08-01`;
    console.log("[EntraToken] Using App Service MSI endpoint…");
    res = await fetch(url, {
      headers: { "X-IDENTITY-HEADER": identityHeader },
      signal: AbortSignal.timeout(15_000),
    });
  } else {
    // ── VM IMDS fallback ──────────────────────────────────────────────────────
    console.log("[EntraToken] IDENTITY_ENDPOINT not set — using VM IMDS endpoint…");
    res = await fetch(VM_IMDS_URL, {
      headers: { Metadata: "true" },
      signal: AbortSignal.timeout(15_000),
    });
  }

  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(
      `[EntraToken] Token fetch failed: HTTP ${res.status} — ${body}\n` +
      `  IDENTITY_ENDPOINT = ${identityEndpoint ?? "(not set)"}\n` +
      `  IDENTITY_HEADER   = ${identityHeader   ? "(present)" : "(not set)"}`
    );
  }

  const data = (await res.json()) as {
    access_token: string;
    expires_on:   string; // unix seconds as string
  };

  if (!data.access_token) {
    throw new Error("[EntraToken] Response had no access_token: " + JSON.stringify(data).slice(0, 200));
  }

  tokenCache = {
    token:     data.access_token,
    expiresAt: parseInt(data.expires_on, 10) * 1000,
  };

  const expiresInMin = Math.round((tokenCache.expiresAt - now) / 60_000);
  console.log(`[EntraToken] ✅ Token fetched. Expires in ~${expiresInMin} min.`);

  return tokenCache.token;
}

/**
 * Fetch an Entra ID Bearer token for Azure Cognitive Services / Azure OpenAI.
 * Scope: https://cognitiveservices.azure.com/
 * Uses App Service MSI endpoint when available, falls back to VM IMDS.
 */
export async function getAzureOpenAIToken(): Promise<string> {
  const now = Date.now();

  if (cogTokenCache && cogTokenCache.expiresAt - now > REFRESH_BUFFER_MS) {
    return cogTokenCache.token;
  }

  const identityEndpoint = process.env.IDENTITY_ENDPOINT;
  const identityHeader   = process.env.IDENTITY_HEADER;

  let res: Response;

  if (identityEndpoint && identityHeader) {
    const url = `${identityEndpoint}?resource=${encodeURIComponent(COG_RESOURCE)}&api-version=2019-08-01`;
    console.log("[EntraToken/OpenAI] Using App Service MSI endpoint for Cognitive Services…");
    res = await fetch(url, {
      headers: { "X-IDENTITY-HEADER": identityHeader },
      signal: AbortSignal.timeout(15_000),
    });
  } else {
    const vmUrl =
      "http://169.254.169.254/metadata/identity/oauth2/token" +
      "?api-version=2019-08-01" +
      `&resource=${encodeURIComponent(COG_RESOURCE)}`;
    console.log("[EntraToken/OpenAI] IDENTITY_ENDPOINT not set — using VM IMDS for Cognitive Services…");
    res = await fetch(vmUrl, {
      headers: { Metadata: "true" },
      signal: AbortSignal.timeout(15_000),
    });
  }

  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(
      `[EntraToken/OpenAI] Token fetch failed: HTTP ${res.status} — ${body}\n` +
      `  IDENTITY_ENDPOINT = ${identityEndpoint ?? "(not set)"}\n` +
      `  IDENTITY_HEADER   = ${identityHeader   ? "(present)" : "(not set)"}`
    );
  }

  const data = (await res.json()) as {
    access_token: string;
    expires_on:   string;
  };

  if (!data.access_token) {
    throw new Error("[EntraToken/OpenAI] Response had no access_token: " + JSON.stringify(data).slice(0, 200));
  }

  cogTokenCache = {
    token:     data.access_token,
    expiresAt: parseInt(data.expires_on, 10) * 1000,
  };

  const expiresInMin = Math.round((cogTokenCache.expiresAt - now) / 60_000);
  console.log(`[EntraToken/OpenAI] ✅ Cognitive Services token fetched. Expires in ~${expiresInMin} min.`);

  return cogTokenCache.token;
}

/** For the admin status panel — returns cache state without fetching */
export function getTokenStatus(): {
  cached:       boolean;
  expiresInMs?: number;
  expiresInMin?: number;
} {
  if (!tokenCache) return { cached: false };
  const expiresInMs = tokenCache.expiresAt - Date.now();
  return { cached: true, expiresInMs, expiresInMin: Math.round(expiresInMs / 60_000) };
}

File: ledgerlm/ledgerlm/python_backend/utils/entra_token.py
"""
Azure Entra ID token fetcher for PostgreSQL Managed Identity auth.

Azure exposes two different MSI endpoints depending on the host:

  App Service / Container Apps:
    URL    → $IDENTITY_ENDPOINT?resource=...&api-version=2019-08-01
    Header → X-IDENTITY-HEADER: $IDENTITY_HEADER

  Bare VM / VMSS:
    URL    → http://169.254.169.254/metadata/identity/oauth2/token?...
    Header → Metadata: true

We try the App Service path first and fall back to VM IMDS.
Token is cached in-process and refreshed 5 minutes before expiry.
"""

import os
import time
import logging
import requests

logger = logging.getLogger(__name__)

PG_RESOURCE = "https://ossrdbms-aad.database.windows.net"

# VM IMDS fallback — NOT available inside App Service containers
VM_IMDS_URL = (
    "http://169.254.169.254/metadata/identity/oauth2/token"
    "?api-version=2019-08-01"
    f"&resource={PG_RESOURCE}"
)

REFRESH_BUFFER_SECONDS = 5 * 60  # refresh 5 min before expiry

_token_cache: dict = {}  # keys: token, expires_at


def get_entra_token() -> str:
    """Fetch (or return cached) Azure Entra access token for Postgres auth."""
    now = time.time()
    expires_at = _token_cache.get("expires_at", 0)

    if _token_cache.get("token") and expires_at - now > REFRESH_BUFFER_SECONDS:
        return _token_cache["token"]

    identity_endpoint = os.environ.get("IDENTITY_ENDPOINT")
    identity_header   = os.environ.get("IDENTITY_HEADER")

    try:
        if identity_endpoint and identity_header:
            # ── Azure App Service / Container Apps MSI ────────────────────────
            url = f"{identity_endpoint}?resource={PG_RESOURCE}&api-version=2019-08-01"
            logger.info("[EntraToken] Fetching token via App Service IDENTITY_ENDPOINT...")
            resp = requests.get(
                url,
                headers={"X-IDENTITY-HEADER": identity_header},
                timeout=15,
            )
        else:
            # ── VM IMDS fallback ──────────────────────────────────────────────
            logger.info("[EntraToken] IDENTITY_ENDPOINT not set — using VM IMDS...")
            resp = requests.get(
                VM_IMDS_URL,
                headers={"Metadata": "true"},
                timeout=15,
            )

        resp.raise_for_status()

    except requests.RequestException as exc:
        raise RuntimeError(
            f"[EntraToken] Token fetch failed: {exc}\n"
            f"  IDENTITY_ENDPOINT = {identity_endpoint or '(not set)'}\n"
            f"  IDENTITY_HEADER   = {'(present)' if identity_header else '(not set)'}"
        ) from exc

    data = resp.json()
    if "access_token" not in data:
        raise RuntimeError(
            f"[EntraToken] No access_token in response: {str(data)[:200]}"
        )

    _token_cache["token"]      = data["access_token"]
    _token_cache["expires_at"] = int(data["expires_on"])  # unix seconds

    expires_in_min = round((_token_cache["expires_at"] - now) / 60)
    logger.info(f"[EntraToken] Token fetched via App Service MSI. Expires in ~{expires_in_min} minutes.")

    return _token_cache["token"]


def get_token_status() -> dict:
    """Return cache state without fetching — for health/status endpoints."""
    if not _token_cache.get("token"):
        return {"cached": False}
    expires_in_s = _token_cache["expires_at"] - time.time()
    return {
        "cached": True,
        "expires_in_seconds": round(expires_in_s),
        "expires_in_minutes": round(expires_in_s / 60),
    }
